<a href="https://colab.research.google.com/github/yujinc129-oss/Drama_data_project/blob/main/notebooks/01_merge_pre.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [231]:
import pandas as pd
import numpy as np

# 배우 데이터셋 만들기

## 원본 데이터 확인

In [232]:
actors_2 = pd.read_csv('https://github.com/yujinc129-oss/Drama_data_project/raw/refs/heads/main/data/raw/actors_info/mydramalistactors_tocsv.csv', header=0)
actors_3 = pd.read_csv('https://github.com/yujinc129-oss/Drama_data_project/raw/refs/heads/main/data/raw/actors_info/wiki_actors.csv')
actors_2.head()

person  ranking  likes_num   nationality gender            dob  \
0  Lee Min Ho        1      13520  South Korean   Male  June 22, 1987   
1  Lee Min Ho        1      13520  South Korean   Male  June 22, 1987   
2  Lee Min Ho        1      13520  South Korean   Male  June 22, 1987   
3  Lee Min Ho        1      13520  South Korean   Male  June 22, 1987   
4  Lee Min Ho        1      13520  South Korean   Male  June 22, 1987   

                        title  \
0               Ask the Stars   
1                    Pachinko   
2   The King: Eternal Monarch   
3              7 First Kisses   
4  The Legend of the Blue Sea   

                                          title_info  title_ratings  
0  Korean Drama, 2023, 0 eps\n\n\nGong Ryong\n\n(...            NaN  
1  Korean Drama, 2022, 8 eps\n\n\nKo Han Su\n\n(M...            8.4  
2  Korean Drama, 2020, 16 eps\n\n\nLee Gon\n\n(Ma...            8.1  
3  Korean Drama, 2016, 8 eps\n\n\nLee Min Ho\n\n(...            6.9  
4  Korean Drama, 2016, 20 eps\n\n\nHeo Joon Jae /...            8.4

In [233]:
len(actors_2[actors_2.nationality == 'South Korean'].person.unique())

925

In [234]:
actors_3.head()

actor_id    actor_name       drama_name  \
0  96612cd9-183e-4530-9e0d-d85322e6ec4c  Jang Do Yoon    Sing My Crush   
1  96612cd9-183e-4530-9e0d-d85322e6ec4c  Jang Do Yoon     Finland Papa   
2  96612cd9-183e-4530-9e0d-d85322e6ec4c  Jang Do Yoon    My Sweet Dear   
3  96612cd9-183e-4530-9e0d-d85322e6ec4c  Jang Do Yoon  Lemon Car Video   
4  b5f2eca0-6e6b-474d-88ad-46f4feaa0a8e  Son Hyun Woo    Sing My Crush   

         character_name          role  
0            Han Ba Ram     Main Role  
1     Toto / Seung Yoon  Support Role  
2                Ye Jun  Support Role  
3  Jang Do Yoon (Ep. 1)     Main Role  
4            Im Han Tae     Main Role

##서로 얼마나 일치하는지 확인
세 개 다 합치면 766개, 두 개 합치면 1712개.

일단 둘 다 돌려서 확인해보기.

생년은 촬영 당시 나이로 환산하기



In [235]:
ac_3 = actors_3.actor_name.copy()
ac_3_pre = ac_3.str.lower().str.replace(' ','')

In [236]:
ac_2 = actors_2[actors_2.nationality == 'South Korean'].person.copy()
ac_2_pre = ac_2.str.lower().str.replace(' ','')

##데이터 정제_1

병합을 위해 actor, drama 컬럼 소문자 변환, 공백 제거

In [237]:
#ac_2
ac_2 = actors_2[actors_2.nationality =='South Korean'][['person','gender','dob']].copy()
ac_2.person = ac_2.person.str.lower().str.replace(' ','')
ac_2 = ac_2.rename(columns={'person':'actor'})
ac_2 = ac_2.drop_duplicates()

In [238]:
#ac_3
ac_3 = actors_3[['actor_name','drama_name','role']].copy()
ac_3.actor_name = ac_3.actor_name.str.lower().str.replace(' ','')
ac_3.drama_name= ac_3.drama_name.str.lower().str.replace(' ','')
ac_3 = ac_3.rename(columns={'actor_name':'actor','drama_name':'drama'})
ac_3

actor               drama          role
0      jangdoyoon         singmycrush     Main Role
1      jangdoyoon         finlandpapa  Support Role
2      jangdoyoon         mysweetdear  Support Role
3      jangdoyoon       lemoncarvideo     Main Role
4      sonhyunwoo         singmycrush     Main Role
...           ...                 ...           ...
8654  leemoonjung        hearttoheart  Support Role
8655   kimyoungae       killme,healme  Support Role
8656   parktamhee  enchantingneighbor     Main Role
8657   shinminsoo         ironladycha     Main Role
8658    ohsangjin   thefamilyiscoming     Main Role

[8659 rows x 3 columns]

##데이터 병합(ac_2 + ac_3)

각 데이터의 공통 컬럼 actor를 기준으로 inner join

데이터 8659개 -> 3969로 축소

In [239]:
actor= pd.merge(ac_3, ac_2, how='inner', on='actor')
actor

actor                 drama          role  gender  \
0      jangdoyoon           singmycrush     Main Role    Male   
1      jangdoyoon           finlandpapa  Support Role    Male   
2      jangdoyoon           mysweetdear  Support Role    Male   
3      jangdoyoon         lemoncarvideo     Main Role    Male   
4       junghaein           d.p.season2     Main Role    Male   
...           ...                   ...           ...     ...   
3964  parkyoochun  thegirlwhoseesscents     Main Role    Male   
3965      kimjoon          cityofthesun     Main Role    Male   
3966         jayb           dreamknight     Main Role    Male   
3967    minhyorin    persevere,goohaera     Main Role  Female   
3968   kimyoungae         killme,healme  Support Role  Female   

                    dob  
0       August  6, 1995  
1       August  6, 1995  
2       August  6, 1995  
3       August  6, 1995  
4        April  1, 1988  
...                 ...  
3964      June  4, 1986  
3965  February  3, 1984  
3966   January  6, 1994  
3967  February 28, 1986  
3968     April 21, 1951  

[3969 rows x 5 columns]

In [240]:
len(actor.actor.unique())

814

##데이터 정제_2 (연도 추출)

배우 생년 추출

In [241]:
#태어난 연도만 뽑기
actor.dob = actor.dob.str.slice(start=-4)
actor = actor.rename(columns={'dob' : 'birth_y'})

In [242]:
actor

actor                 drama          role  gender birth_y
0      jangdoyoon           singmycrush     Main Role    Male    1995
1      jangdoyoon           finlandpapa  Support Role    Male    1995
2      jangdoyoon           mysweetdear  Support Role    Male    1995
3      jangdoyoon         lemoncarvideo     Main Role    Male    1995
4       junghaein           d.p.season2     Main Role    Male    1988
...           ...                   ...           ...     ...     ...
3964  parkyoochun  thegirlwhoseesscents     Main Role    Male    1986
3965      kimjoon          cityofthesun     Main Role    Male    1984
3966         jayb           dreamknight     Main Role    Male    1994
3967    minhyorin    persevere,goohaera     Main Role  Female    1986
3968   kimyoungae         killme,healme  Support Role  Female    1951

[3969 rows x 5 columns]

#드라마 데이터셋 만들기

##원본 데이터 확인
뭔가 1000개 넘개 데이터가 사라졌지만...
일단 레쓰고

나중에 더 보완할만한 데이터 있으면 찾아서 적용해보자


In [243]:
drama_data = pd.read_csv('https://github.com/yujinc129-oss/Drama_data_project/raw/refs/heads/main/data/raw/drama_info/kdramalist.csv')

In [244]:
drama_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1278 entries, 0 to 1277
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   drama_name        1278 non-null   object 
 1   Genres            1277 non-null   object 
 2   Tags              1278 non-null   object 
 3   Episodes          1278 non-null   int64  
 4   start airing      1255 non-null   object 
 5   end airing        1221 non-null   object 
 6   Aired On          1222 non-null   object 
 7   Original Network  1250 non-null   object 
 8   Duration          1230 non-null   float64
 9   score             1239 non-null   float64
 10  scored by         1239 non-null   float64
 11  Ranked            1278 non-null   int64  
 12  Popularity        1278 non-null   int64  
 13  Content Rating    1278 non-null   object 
 14  Watchers          1278 non-null   object 
 15  actors            1278 non-null   object 
 16  platforms         1278 non-null   object 


In [245]:
drama_data.head()

drama_name                                        Genres  \
0    100 Days My Prince  ['historical', 'comedy', 'romance', 'drama']   
1      12 Signs of Love                 ['comedy', 'romance', 'life']   
2      12 Years Promise        ['food', 'romance', 'drama', 'family']   
3              18 Again       ['romance', 'life', 'drama', 'fantasy']   
4  365: Repeat the Year   ['thriller', 'mystery', 'drama', 'fantasy']   

                                                Tags  Episodes start airing  \
0  ['amnesia', 'hidden identity', 'marriage of co...        16    10-Sep-18   
1                            ['writer female lead ']        16    15-Feb-12   
2  ['unexpected pregnancy', 'miscarriage', 'teena...        26    22-Mar-14   
3  ['second chance', 'first love', 'father-son re...        16    21-Sep-20   
4  ['strong female lead', 'time travel', 'nice ma...        24    23-Mar-20   

  end airing                   Aired On Original Network  Duration  score  \
0  30-Oct-18      ['tuesday', 'monday']          ['tvN']      75.0    8.3   
1   5-Apr-12  ['wednesday', 'thursday']          ['tvN']      65.0    7.1   
2  29-Jun-14     ['saturday', 'sunday']         ['jTBC']      70.0    7.2   
3  10-Nov-20      ['monday', 'tuesday']         ['jTBC']      70.0    8.7   
4  28-Apr-20      ['monday', 'tuesday']          ['MBC']      35.0    8.6   

   scored by  Ranked  Popularity           Content Rating Watchers  \
0    21274.0     554         136  15+ - Teens 15 or older   45,852   
1      893.0    5506        3327            Not Yet Rated    2,543   
2     2716.0    5395        1783  15+ - Teens 15 or older    5,435   
3    21409.0     105         125  15+ - Teens 15 or older   47,527   
4     7598.0     250         438  15+ - Teens 15 or older   20,963   

                                              actors  \
0  ['Doh Kyung Soo', 'Nam Ji Hyun', 'Kim Seon Ho'...   
1  ['Ohn Joo Wan', 'Yoon Jin Seo', 'Lee Yong Woo'...   
2  ['Lee So Yeon', 'Namkoong Min', 'Lee Tae Im', ...   
3  ['Kim Ha Neul', 'Yoon Sang Hyun', 'Lee Do Hyun...   
4  ['Lee Joon Hyuk', 'Nam Ji Hyun', 'Kim Jee Soo'...   

                                           platforms  imdb_rating  \
0  ['WeTV', 'Netflix', 'K-DRAMA', 'Apple TV', 'Vi...          7.7   
1                                                 []          NaN   
2                                           ['Viki']          7.3   
3                                           ['WeTV']          8.2   
4                                           ['Viki']          8.0   

  imdb_user_count                                   imdb_description  
0           2,188  Upon losing his memory, a crown prince encount...  
1             NaN                                                NaN  
2             181  A pregnant teen is forced by her family to lea...  
3           1,847  A 37-year-old man on the verge of being divorc...  
4             656  A story where ten people get the chance to go ...

In [246]:
drama_data.columns

Index(['drama_name', 'Genres', 'Tags', 'Episodes', 'start airing',
       'end airing', 'Aired On', 'Original Network', 'Duration', 'score',
       'scored by', 'Ranked', 'Popularity', 'Content Rating', 'Watchers',
       'actors', 'platforms', 'imdb_rating', 'imdb_user_count',
       'imdb_description'],
      dtype='object')

In [247]:
drama = drama_data[['drama_name','Genres','start airing','Aired On','Original Network','score','scored by',]].copy()

##데이터 정제
방영연도 추출

genre, aired on, original network 문자열 제거

score, scored_by 로 가중평점 컬럼 생성

드라마 이름 소문자, 공백 제거 for 병합

In [248]:
drama.Genres.info()

<class 'pandas.core.series.Series'>
RangeIndex: 1278 entries, 0 to 1277
Series name: Genres
Non-Null Count  Dtype 
--------------  ----- 
1277 non-null   object
dtypes: object(1)
memory usage: 10.1+ KB


In [249]:
drama.dropna(inplace=True)

In [250]:
drama

drama_name                                           Genres  \
0       100 Days My Prince     ['historical', 'comedy', 'romance', 'drama']   
1         12 Signs of Love                    ['comedy', 'romance', 'life']   
2         12 Years Promise           ['food', 'romance', 'drama', 'family']   
3                 18 Again          ['romance', 'life', 'drama', 'fantasy']   
4     365: Repeat the Year      ['thriller', 'mystery', 'drama', 'fantasy']   
...                    ...                                              ...   
1272             Your Lady      ['romance', 'drama', 'family', 'melodrama']   
1273  Your Neighbor's Wife       ['mystery', 'comedy', 'romance', 'family']   
1275          Youth of May      ['historical', 'romance', 'youth', 'drama']   
1276          Yumi's Cells  ['psychological', 'comedy', 'romance', 'drama']   
1277      Zombie Detective     ['thriller', 'mystery', 'comedy', 'fantasy']   

     start airing                                           Aired On  \
0       10-Sep-18                              ['tuesday', 'monday']   
1       15-Feb-12                          ['wednesday', 'thursday']   
2       22-Mar-14                             ['saturday', 'sunday']   
3       21-Sep-20                              ['monday', 'tuesday']   
4       23-Mar-20                              ['monday', 'tuesday']   
...           ...                                                ...   
1272    18-Feb-13  ['monday', 'tuesday', 'wednesday', 'thursday',...   
1273    14-Oct-13                              ['monday', 'tuesday']   
1275     3-May-21                              ['monday', 'tuesday']   
1276    17-Sep-21                             ['friday', 'saturday']   
1277    21-Sep-20                              ['monday', 'tuesday']   

     Original Network  score  scored by  
0             ['tvN']    8.3    21274.0  
1             ['tvN']    7.1      893.0  
2            ['jTBC']    7.2     2716.0  
3            ['jTBC']    8.7    21409.0  
4             ['MBC']    8.6     7598.0  
...               ...    ...        ...  
1272          ['SBS']    7.2       51.0  
1273         ['jTBC']    7.3       65.0  
1275         ['KBS2']    8.8    12576.0  
1276          ['tvN']    8.5    13469.0  
1277         ['KBS2']    8.2     6840.0  

[1202 rows x 7 columns]

In [251]:
#연도 변환
date_origin = drama['start airing']
date_clean = pd.to_datetime(date_origin, errors='coerce')
year = date_clean.dt.year

drama['start airing'] = year

/tmp/ipython-input-332174941.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  date_clean = pd.to_datetime(date_origin, errors='coerce')


In [252]:
drama

drama_name                                           Genres  \
0       100 Days My Prince     ['historical', 'comedy', 'romance', 'drama']   
1         12 Signs of Love                    ['comedy', 'romance', 'life']   
2         12 Years Promise           ['food', 'romance', 'drama', 'family']   
3                 18 Again          ['romance', 'life', 'drama', 'fantasy']   
4     365: Repeat the Year      ['thriller', 'mystery', 'drama', 'fantasy']   
...                    ...                                              ...   
1272             Your Lady      ['romance', 'drama', 'family', 'melodrama']   
1273  Your Neighbor's Wife       ['mystery', 'comedy', 'romance', 'family']   
1275          Youth of May      ['historical', 'romance', 'youth', 'drama']   
1276          Yumi's Cells  ['psychological', 'comedy', 'romance', 'drama']   
1277      Zombie Detective     ['thriller', 'mystery', 'comedy', 'fantasy']   

      start airing                                           Aired On  \
0             2018                              ['tuesday', 'monday']   
1             2012                          ['wednesday', 'thursday']   
2             2014                             ['saturday', 'sunday']   
3             2020                              ['monday', 'tuesday']   
4             2020                              ['monday', 'tuesday']   
...            ...                                                ...   
1272          2013  ['monday', 'tuesday', 'wednesday', 'thursday',...   
1273          2013                              ['monday', 'tuesday']   
1275          2021                              ['monday', 'tuesday']   
1276          2021                             ['friday', 'saturday']   
1277          2020                              ['monday', 'tuesday']   

     Original Network  score  scored by  
0             ['tvN']    8.3    21274.0  
1             ['tvN']    7.1      893.0  
2            ['jTBC']    7.2     2716.0  
3            ['jTBC']    8.7    21409.0  
4             ['MBC']    8.6     7598.0  
...               ...    ...        ...  
1272          ['SBS']    7.2       51.0  
1273         ['jTBC']    7.3       65.0  
1275         ['KBS2']    8.8    12576.0  
1276          ['tvN']    8.5    13469.0  
1277         ['KBS2']    8.2     6840.0  

[1202 rows x 7 columns]

In [253]:
#문자열 제거
drama['Aired On'] = drama['Aired On'].str.slice(1, -1).str.replace("'","").str.split(',')
drama['Original Network']=drama['Original Network'].str.slice(1,-1).str.replace("'","").str.upper().str.split(',')
drama['Genres']= drama['Genres'].str.slice(1,-1).str.replace("'",'').str.split(',')

In [254]:
#가중 평점 변환
drama['scored by'].describe()

count     1202.000000
mean      7144.923461
std      12331.322010
min          1.000000
25%        354.250000
50%       1982.500000
75%       8104.000000
max      92027.000000
Name: scored by, dtype: float64

In [255]:
drama['score'].describe()

count    1202.000000
mean        7.703411
std         0.591429
min         4.000000
25%         7.300000
50%         7.700000
75%         8.100000
max        10.000000
Name: score, dtype: float64

In [256]:
m = 8500
drama['score_weighted'] = (drama['scored by'] / (drama['scored by'] + m)) * drama['score'] + (m / (drama['scored by']+m)) * drama['score'].mean()
drama.drop(columns=['score','scored by'], inplace=True)
drama.score_weighted=drama.score_weighted.round(2)
drama

drama_name                                      Genres  \
0       100 Days My Prince     [historical,  comedy,  romance,  drama]   
1         12 Signs of Love                   [comedy,  romance,  life]   
2         12 Years Promise           [food,  romance,  drama,  family]   
3                 18 Again          [romance,  life,  drama,  fantasy]   
4     365: Repeat the Year      [thriller,  mystery,  drama,  fantasy]   
...                    ...                                         ...   
1272             Your Lady      [romance,  drama,  family,  melodrama]   
1273  Your Neighbor's Wife       [mystery,  comedy,  romance,  family]   
1275          Youth of May      [historical,  romance,  youth,  drama]   
1276          Yumi's Cells  [psychological,  comedy,  romance,  drama]   
1277      Zombie Detective     [thriller,  mystery,  comedy,  fantasy]   

      start airing                                           Aired On  \
0             2018                                 [tuesday,  monday]   
1             2012                             [wednesday,  thursday]   
2             2014                                [saturday,  sunday]   
3             2020                                 [monday,  tuesday]   
4             2020                                 [monday,  tuesday]   
...            ...                                                ...   
1272          2013  [monday,  tuesday,  wednesday,  thursday,  fri...   
1273          2013                                 [monday,  tuesday]   
1275          2021                                 [monday,  tuesday]   
1276          2021                                [friday,  saturday]   
1277          2020                                 [monday,  tuesday]   

     Original Network  score_weighted  
0               [TVN]            8.13  
1               [TVN]            7.65  
2              [JTBC]            7.58  
3              [JTBC]            8.42  
4               [MBC]            8.13  
...               ...             ...  
1272            [SBS]            7.70  
1273           [JTBC]            7.70  
1275           [KBS2]            8.36  
1276            [TVN]            8.19  
1277           [KBS2]            7.92  

[1202 rows x 6 columns]

In [259]:
#드라마 이름 소문자, 공백 제거
drama.drama_name = drama.drama_name.str.lower().str.replace(' ','')

In [260]:
drama

drama_name                                      Genres  \
0        100daysmyprince     [historical,  comedy,  romance,  drama]   
1          12signsoflove                   [comedy,  romance,  life]   
2         12yearspromise           [food,  romance,  drama,  family]   
3                18again          [romance,  life,  drama,  fantasy]   
4      365:repeattheyear      [thriller,  mystery,  drama,  fantasy]   
...                  ...                                         ...   
1272            yourlady      [romance,  drama,  family,  melodrama]   
1273  yourneighbor'swife       [mystery,  comedy,  romance,  family]   
1275          youthofmay      [historical,  romance,  youth,  drama]   
1276         yumi'scells  [psychological,  comedy,  romance,  drama]   
1277     zombiedetective     [thriller,  mystery,  comedy,  fantasy]   

      start airing                                           Aired On  \
0             2018                                 [tuesday,  monday]   
1             2012                             [wednesday,  thursday]   
2             2014                                [saturday,  sunday]   
3             2020                                 [monday,  tuesday]   
4             2020                                 [monday,  tuesday]   
...            ...                                                ...   
1272          2013  [monday,  tuesday,  wednesday,  thursday,  fri...   
1273          2013                                 [monday,  tuesday]   
1275          2021                                 [monday,  tuesday]   
1276          2021                                [friday,  saturday]   
1277          2020                                 [monday,  tuesday]   

     Original Network  score_weighted  
0               [TVN]            8.13  
1               [TVN]            7.65  
2              [JTBC]            7.58  
3              [JTBC]            8.42  
4               [MBC]            8.13  
...               ...             ...  
1272            [SBS]            7.70  
1273           [JTBC]            7.70  
1275           [KBS2]            8.36  
1276            [TVN]            8.19  
1277           [KBS2]            7.92  

[1202 rows x 6 columns]

#배우 + 드라마 데이터 병합

드라마 이름 기준으로 inner join

In [261]:
merged_data= pd.merge(actor, drama, left_on='drama', right_on='drama_name')
merged_data

actor                   drama          role  gender birth_y  \
0         junghaein                snowdrop     Main Role    Male    1988   
1         junghaein                    d.p.     Main Role    Male    1988   
2         junghaein        apieceofyourmind     Main Role    Male    1988   
3         junghaein          onespringnight     Main Role    Male    1988   
4         junghaein      somethingintherain     Main Role    Male    1988   
...             ...                     ...           ...     ...     ...   
2212    kimjongkook            theproducers  Support Role    Male    1976   
2213    jungyoonhak  myunfortunateboyfriend     Main Role    Male    1984   
2214  jungjoonyoung                thelover     Main Role    Male    1989   
2215        kimjoon            cityofthesun     Main Role    Male    1984   
2216     kimyoungae           killme,healme  Support Role  Female    1951   

                  drama_name                                      Genres  \
0                   snowdrop   [romance,  drama,  melodrama,  political]   
1                       d.p.                 [action,  military,  drama]   
2           apieceofyourmind        [romance,  life,  drama,  melodrama]   
3             onespringnight        [romance,  life,  drama,  melodrama]   
4         somethingintherain        [romance,  life,  drama,  melodrama]   
...                      ...                                         ...   
2212            theproducers                  [comedy,  romance,  drama]   
2213  myunfortunateboyfriend                           [romance,  drama]   
2214                thelover                   [comedy,  romance,  life]   
2215            cityofthesun         [mystery,  law,  drama,  melodrama]   
2216           killme,healme  [psychological,  comedy,  romance,  drama]   

      start airing                Aired On  Original Network  score_weighted  
0             2021     [saturday,  sunday]            [JTBC]            8.20  
1             2021                [friday]         [NETFLIX]            8.43  
2             2020      [monday,  tuesday]             [TVN]            7.97  
3             2019  [thursday,  wednesday]   [MBC,  NETFLIX]            8.17  
4             2018     [friday,  saturday]  [JTBC,  NETFLIX]            7.92  
...            ...                     ...               ...             ...  
2212          2015     [friday,  saturday]            [KBS2]            7.64  
2213          2015     [friday,  saturday]    [MBC DRAMANET]            7.54  
2214          2015              [thursday]            [MNET]            7.80  
2215          2015     [friday,  saturday]    [MBC DRAMANET]            7.70  
2216          2015  [wednesday,  thursday]             [MBC]            8.54  

[2217 rows x 11 columns]

##데이터 정제_1

start airing 과 birth_y 컬럼으로 방영 당시 배우 나이 컬럼 생성

In [262]:
merged_data['birth_y'] = pd.to_numeric(merged_data['birth_y'], errors='coerce')
merged_data['age'] = merged_data['start airing'] - merged_data['birth_y'] + 1

In [263]:
first_data = merged_data.drop(columns=['birth_y','drama_name'])

In [264]:
first_data

actor                   drama          role  gender  \
0         junghaein                snowdrop     Main Role    Male   
1         junghaein                    d.p.     Main Role    Male   
2         junghaein        apieceofyourmind     Main Role    Male   
3         junghaein          onespringnight     Main Role    Male   
4         junghaein      somethingintherain     Main Role    Male   
...             ...                     ...           ...     ...   
2212    kimjongkook            theproducers  Support Role    Male   
2213    jungyoonhak  myunfortunateboyfriend     Main Role    Male   
2214  jungjoonyoung                thelover     Main Role    Male   
2215        kimjoon            cityofthesun     Main Role    Male   
2216     kimyoungae           killme,healme  Support Role  Female   

                                          Genres  start airing  \
0      [romance,  drama,  melodrama,  political]          2021   
1                    [action,  military,  drama]          2021   
2           [romance,  life,  drama,  melodrama]          2020   
3           [romance,  life,  drama,  melodrama]          2019   
4           [romance,  life,  drama,  melodrama]          2018   
...                                          ...           ...   
2212                  [comedy,  romance,  drama]          2015   
2213                           [romance,  drama]          2015   
2214                   [comedy,  romance,  life]          2015   
2215         [mystery,  law,  drama,  melodrama]          2015   
2216  [psychological,  comedy,  romance,  drama]          2015   

                    Aired On  Original Network  score_weighted  age  
0        [saturday,  sunday]            [JTBC]            8.20   34  
1                   [friday]         [NETFLIX]            8.43   34  
2         [monday,  tuesday]             [TVN]            7.97   33  
3     [thursday,  wednesday]   [MBC,  NETFLIX]            8.17   32  
4        [friday,  saturday]  [JTBC,  NETFLIX]            7.92   31  
...                      ...               ...             ...  ...  
2212     [friday,  saturday]            [KBS2]            7.64   40  
2213     [friday,  saturday]    [MBC DRAMANET]            7.54   32  
2214              [thursday]            [MNET]            7.80   27  
2215     [friday,  saturday]    [MBC DRAMANET]            7.70   32  
2216  [wednesday,  thursday]             [MBC]            8.54   65  

[2217 rows x 10 columns]

##데이터 정제_2
장르 재 카테고리화

In [265]:
def transform_genre(genre_list):
    """
    장르 리스트를 새로운 카테고리로 변환하는 함수 (공백 제거 기능 추가)
    """
    genre_map = {
        'thriller': ['mystery', 'thriller', 'crime', 'psychological', 'horror', 'tokusatsu'],
        'sf': ['fantasy', 'sci-fi', 'supernatural', 'adventure'],
        'society': ['business', 'law', 'political', 'medical'],
        'youth': ['youth', 'school', 'friendship'],
        'hist_war': ['military', 'war', 'historical'],
        'comedy': ['comedy', 'sitcom'],
        'romance': ['romance', 'melodrama'],
        'action': ['action', 'martial arts'],
        'drama': ['drama', 'life', 'family'],
        'etc': ['mature','food', 'music', 'sports']
    }

    transformed_genres = set()


    for genre in genre_list:
        # ✨ 이 부분이 핵심! 각 장르의 앞뒤 공백을 제거합니다.
        genre = genre.strip()

        for category, values in genre_map.items():
            if genre in values:
                transformed_genres.add(category)
                break

    return list(transformed_genres)

# --- 사용자 데이터로 테스트 ---
genres0 = ['romance', ' drama', ' melodrama', ' political']

# --- 함수 실행 및 결과 확인 ---
# sorted() 함수를 추가하여 결과를 항상 동일한 순서로 볼 수 있게 합니다.
transformed = sorted(transform_genre(genres0))
print(f"Original: {genres0} -> Transformed: {transformed}")

Original: ['romance', ' drama', ' melodrama', ' political'] -> Transformed: ['drama', 'romance', 'society']


In [266]:
for x in range(len(first_data['Genres'])):
  first_data['Genres'].loc[x] = transform_genre(first_data['Genres'].loc[x])


스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  first_data['Genres'].loc[x] = transform_genre(first_data['Genres'].loc[x])
/tmp/ipython-input-3712344801.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  first_data['Genres'].loc[x] = transform_genre(first_data['Genres'].loc[x])
/tmp/ipython-input-3712344801.py:2: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are set

In [267]:
first_data

actor                   drama          role  gender  \
0         junghaein                snowdrop     Main Role    Male   
1         junghaein                    d.p.     Main Role    Male   
2         junghaein        apieceofyourmind     Main Role    Male   
3         junghaein          onespringnight     Main Role    Male   
4         junghaein      somethingintherain     Main Role    Male   
...             ...                     ...           ...     ...   
2212    kimjongkook            theproducers  Support Role    Male   
2213    jungyoonhak  myunfortunateboyfriend     Main Role    Male   
2214  jungjoonyoung                thelover     Main Role    Male   
2215        kimjoon            cityofthesun     Main Role    Male   
2216     kimyoungae           killme,healme  Support Role  Female   

                                   Genres  start airing  \
0               [drama, society, romance]          2021   
1               [drama, action, hist_war]          2021   
2                        [drama, romance]          2020   
3                        [drama, romance]          2019   
4                        [drama, romance]          2018   
...                                   ...           ...   
2212             [drama, comedy, romance]          2015   
2213                     [drama, romance]          2015   
2214             [drama, comedy, romance]          2015   
2215  [drama, romance, thriller, society]          2015   
2216   [drama, thriller, comedy, romance]          2015   

                    Aired On  Original Network  score_weighted  age  
0        [saturday,  sunday]            [JTBC]            8.20   34  
1                   [friday]         [NETFLIX]            8.43   34  
2         [monday,  tuesday]             [TVN]            7.97   33  
3     [thursday,  wednesday]   [MBC,  NETFLIX]            8.17   32  
4        [friday,  saturday]  [JTBC,  NETFLIX]            7.92   31  
...                      ...               ...             ...  ...  
2212     [friday,  saturday]            [KBS2]            7.64   40  
2213     [friday,  saturday]    [MBC DRAMANET]            7.54   32  
2214              [thursday]            [MNET]            7.80   27  
2215     [friday,  saturday]    [MBC DRAMANET]            7.70   32  
2216  [wednesday,  thursday]             [MBC]            8.54   65  

[2217 rows x 10 columns]

#추가 데이터 정제

한 드라마 안에 동일 인물이 다른 나이로 두 번 등장하는 것을 파악함.

이를 제거하여 데이터의 퀄리티를 상승.

(by 정현)